wiki/Bazel-User-Guide.md

```diff
@@ -0,0 +1,277 @@
+
+```
+deps = [ "//app",]
+```
+Here, `deps` is used to define the dependencies which is a type of dependencies called `deps dependencies` and it includes the files/directory/target which are dependent. From the above example the dependency is the `app` target which is defined in the [Build file of app package](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L616).
+
+Example of Dependencies
+1. [srcs dependencies](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L617)
+2. [deps dependencies](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L622)
+
+**[Loading an extension](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L13)**
+
+Bazel extensions are files ending in .bzl. Use the load statement to import a symbol from an extension.
+
+```
+load("@io_bazel_rules_kotlin//kotlin:android.bzl", "kt_android_library")
+```
+Here, we are loading `android.bzl` and we are going to use it with a symbol name `kt_android_library`.
+Arguments to the load function must be string literals. load statements must appear at top-level in the file.
+
+**[Visibility of a file target](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L621)**
+
+With the example from our codebase, target `app` whose visibility is public. 
+
+ - `visibility = ["//visibility:public"],` - Anyone can use this target.
+ - `"//visibility:private"` - Only targets in this package can use this target.
+
+**[Testing](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L719)**
+
+when we want to run test cases on Bazel build environment, a test target needs to be set up correctly:
+
+```bazel
+load("//:oppia_android_test.bzl", "oppia_android_test")
+
+oppia_android_test(
+    name = "MathExpressionInteractionsViewTest",
+    srcs = [
+        "MathExpressionInteractionsViewTest.kt",
+        "//app:data_binder_mapper_impl"
+    ],
+    custom_package = "org.oppia.android.app.test",
+    test_manifest = "//app:test_manifest",
+    enable_data_binding = True,
+    test_class = "org.oppia.android.app.customview.interaction.MathExpressionInteractionsViewTest",
+    deps = [
+      ...
+    ],
+)
+```
+
+The above assumes that the corresponding test requires resources. If it doesn't, the definition can be a bit simpler:
+
+```bazel
+load("//:oppia_android_test.bzl", "oppia_android_test")
+
+oppia_android_test(
+    name = "MathExpressionAccessibilityUtilTest",
+    srcs = ["MathExpressionAccessibilityUtilTest.kt"],
+    custom_package = "org.oppia.android.app.utility.math",
+    test_class = "org.oppia.android.app.utility.math.MathExpressionAccessibilityUtilTest",
+    test_manifest = "//app:test_manifest",
+    deps = [
+      ...
+    ],
```

Comment:
Can we make sure the order here is consistent between this block and the one above? Also, they should follow whatever buildifier enforces.

Reply:
Done